packages/loader/container-loader/src/deltaManager.ts

```diff
@@ -1104,14 +1114,13 @@ export class DeltaManager
      * @returns A promise that resolves when the connection is reestablished or we stop trying
      */
     private async reconnectOnError(
-        connection: IDocumentDeltaConnection,
         requestedMode: ConnectionMode,
         error: ICriticalContainerError,
     ) {
         // We quite often get protocol errors before / after observing nack/disconnect
         // we do not want to run through same sequence twice.
-        // We should correctly unregister all event listeners not to re-enter here again!
-        assert(connection === this.connection);
+        // If we're already disconnected/disconnecting it's not appropriate to call this again.
+        assert(this.connection !== undefined);
```

Comment:
This is weaker check than we have before - what if we already established new connection and event fired for old connection?

Reply:
I had this thought too (and I do think we can improve these connect/disconnect flows) but we should be protected here.

We only assign to `this.connection` after asserting it's `undefined`, and only attach the listeners after doing so.  So as long as we clean up any event listeners on old connections when setting `this.connection = undefined`, the number of registered event listeners should always be 0 or 1.  And this is what we do in `disconnectFromDeltaStream`.

So by keeping the register/unregister tightly correlated with defined/undefined, we simultaneously avoid leaking the connection reference itself as well as event listeners upon it.